src/Layer/LabelLayer.js

```diff
@@ -0,0 +1,189 @@
+ * LabelLayer.  Default is true. You should not change this, as it is used
+ * internally for optimisation.
+ * @property {string} crs - The crs of this layer.
+ */
+class LabelLayer extends Layer {
+    constructor(id, crs) {
+        super(id);
+
+        this.isLabelLayer = true;
+        this.crs = crs;
+    }
+
+    /**
+     * Reads each {@link FeatureGeometry} that contains label configuration, and
+     * creates the corresponding {@link Label}. To create a `Label`, a geometry
+     * needs to have a `label` object with at least a few properties:
+     * - `content`, which refers to `Label#content`
+     * - `position`, which refers to `Label#position`
+     * - (optional) `config`, containing miscellaneous configuration for the
+     *   label
+     *
+     * The geometry (or its parent Feature) needs to have a Style set.
+     *
+     * @param {FeatureCollection} data - The FeatureCollection to read the
+     * labels from.
+     * @param {Extent} extent
+     *
+     * @return {Label[]} An array containing all the created labels.
+     */
+    convert(data, extent) {
+        const labels = [];
+
+        const layerField = this.style && this.style.text && this.style.text.field;
+
+        // Converting the extent now is faster for further operation
+        extent.as(data.crs, _extent);
+        coord.crs = data.crs;
+
+        data.features.forEach((f) => {
+            // TODO: add support for LINE and POLYGON
+            if (f.type !== FEATURE_TYPES.POINT) {
+                return;
+            }
+
+            const featureField = f.style && f.style.text.field;
+
+            f.geometry.forEach((g) => {
+                const geometryField = g.properties.style && g.properties.style.text.field;
+
+                if (!geometryField && !featureField && !layerField) {
+                    return;
+                } else {
+                    content = g.properties[geometryField || featureField || layerField];
+                }
+
+                // NOTE: this only works because only POINT is supported, it
+                // needs more work for LINE and POLYGON
+                coord.setFromArray(f.vertices, g.size * g.indices[0].offset);
+                if (f.size == 2) { coord.z = 0; }
+
+                if (this.source.isFileSource && !_extent.isPointInside(coord)) { return; }
+
+                // FIXME: Style management needs improvement, see #1318.
+                // NOTE: the config param is empty, but will be filled in the
+                // next version.
+                const label = new Label(content, coord.clone(), g.properties.style || f.style || this.style);
+                label.layer = label;
+
+                // coord.x = (coord.x / data.scale.x) - data.translation.x;
+                // coord.y = (coord.y / data.scale.y) - data.translation.y;
+                coord.as(this.crs, coord);
+                coord.toVector3(label.position);
+                label.needsToBeMoved = true;
```

Comment:
why don't do this in constructor?

```suggestion
      // need to add layer parameter in Label constructor
      this.coord.as(this.layer.crs, this.position);
      this.needsToBeMoved = true;
```

Reply:
This won't work,  `Coordinates#as()` can't take a `THREE.Vector3` as second argument. But the operation could be done in the constructor indeed, I'm trying it.

Reply:
This is done in the constructor now